naga/src/front/wgsl/lower/mod.rs

```diff
@@ -1464,24 +1464,22 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
                     let mut emitter = Emitter::default();
                     emitter.start(&ctx.function.expressions);
 
-                    let value =
-                        self.expression(l.init, &mut ctx.as_expression(block, &mut emitter))?;
-
-                    // The WGSL spec says that any expression that refers to a
-                    // `let`-bound variable is not a const expression. This
-                    // affects when errors must be reported, so we can't even
-                    // treat suitable `let` bindings as constant as an
-                    // optimization.
-                    ctx.local_expression_kind_tracker.force_non_const(value);
-
                     let explicit_ty = l
                         .ty
                         .map(|ty| self.resolve_ast_type(ty, &mut ctx.as_const(block, &mut emitter)))
                         .transpose()?;
 
-                    if let Some(ty) = explicit_ty {
-                        let mut ctx = ctx.as_expression(block, &mut emitter);
-                        let init_ty = ctx.register_type(value)?;
+                    let mut ectx = ctx.as_expression(block, &mut emitter);
+
+                    let value = if let Some(ty) = explicit_ty {
+                        let (init_ty, lowered_init) = self.type_and_init(
```

Comment:
Shouldn't unconditionally calling `type_and_init()` be okay? If `explicit_ty` is None the initializer will be concretized, which is equivalent to calling `expression()` as your patch does in the else block below.